File: slack_channel/add_karma_reaction_handler.py
import logging
from model.karma import KarmaType

from commands import AddKarmaCommand
from slack_channel.abstract_reaction_handler import AbstractReactionHandler

class AddKarmaReactionHandler(AbstractReactionHandler):

    @property
    def command(self) -> AddKarmaCommand:
        return AddKarmaCommand()
    
    def can_handle(self, reaction_add_event) -> bool:
        reaction = self.parse_reaction(reaction_add_event["reaction"])
        return reaction in AddKarmaCommand.get_karma_reactions()
    
    def _invoke_handler_logic(self, reaction_add_event):
        reactions = AddKarmaCommand.get_karma_reactions()
        try:
            reaction = self.parse_reaction(reaction_add_event["reaction"])
            self.command.execute(awarded_to=reaction_add_event["reactee"],
                                 awarded_by=reaction_add_event["reactor"],
                                 reason="Reacted with " + reaction,
                                 karma_type=reactions[reaction])
            self._send_reaction_response(reaction_add_event)
        except Exception as ex:
            logging.exception(ex)

    def parse_reaction(self, raw_reaction):
        return raw_reaction.split('::')[0]

File: bottomly.net/Bottomly/Slack/ReactionHandlers/AddKarmaReactionHandler.cs
using Bottomly.Commands;
using Bottomly.Repositories;
using Microsoft.Extensions.Logging;
using SlackNet;
using SlackNet.Events;

namespace Bottomly.Slack.ReactionHandlers;

public class AddKarmaReactionHandler(
    AddKarmaCommand command,
    KarmaReactionMap reactionMap,
    IMemberRepository memberRepository,
    ISlackMessageBroker broker,
    ILogger<AddKarmaReactionHandler> logger)
    : IReactionHandler
{
    public bool CanHandle(ReactionAdded reactionEvent)
    {
        var reaction = ParseReaction(reactionEvent.Reaction);
        return reactionMap.IsKarmaReaction(reaction);
    }

    public async Task HandleAsync(ReactionAdded reactionEvent)
    {
        try
        {
            var reaction = ParseReaction(reactionEvent.Reaction);

            var reactor = await memberRepository.GetBySlackIdAsync(reactionEvent.User);
            var reactee = await memberRepository.GetBySlackIdAsync(reactionEvent.ItemUser);

            if (reactor is null || reactee is null)
            {
                logger.LogWarning("Could not resolve reactor ({ReactorId}) or reactee ({ReacteeId})",
                    reactionEvent.User, reactionEvent.ItemUser);
                return;
            }

            await command.ExecuteAsync(
                reactee.Username,
                reactor.Username,
                $"Reacted with {reaction}",
                reactionMap.GetKarmaType(reaction));

            if (reactionEvent.Item is ReactionMessage messageItem)
            {
                await broker.SendReactionAsync("robot_face", messageItem.Channel, messageItem.Ts);
            }
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Error handling reaction {Reaction}", reactionEvent.Reaction);
        }
    }

    private static string ParseReaction(string rawReaction) => rawReaction.Split("::")[0];
}